src/lib/rpc/bestRpcTracker.ts

```diff
@@ -0,0 +1,346 @@
+      timestamp: Date.now(),
+    })
+  );
+}
+
+function isDebugMode() {
+  return localStorage.getItem(JSON.stringify(SHOW_DEBUG_VALUES_KEY)) === "true";
+}
+
+async function probeRpc(chainId: number, provider: Provider, providerUrl: string): Promise<ProbeData> {
+  const controller = new AbortController();
+
+  return await Promise.race([
+    sleep(PROBE_FAIL_TIMEOUT).then(() => {
+      controller.abort();
+      throw new Error("Probe timeout");
+    }),
+    (async function callRpc() {
+      const dataStoreAddress = getContract(chainId, "DataStore");
+      const multicallAddress = getContract(chainId, "Multicall");
+      const dataStore = getDataStoreContract(chainId, provider);
+      const multicall = getMulticallContract(chainId, provider);
+
+      const probeMarketAddress = PROBE_SAMPLE_MARKET[chainId];
+      const probeFieldKey = HASHED_MARKET_CONFIG_KEYS[chainId]?.[probeMarketAddress]?.[PROBE_SAMPLE_FIELD];
+
+      const startTime = Date.now();
+
+      let blockNumber: number | null = null;
+      let isSuccess = false;
+
+      if (!dataStoreAddress || !multicallAddress || !probeMarketAddress || !probeFieldKey || !dataStore || !multicall) {
+        throw new Error("Failed to get RPC probe request data");
+      } else {
+        const dataStoreData = dataStore.interface.encodeFunctionData("getUint", [probeFieldKey]);
+        const multicallData = multicall.interface.encodeFunctionData("blockAndAggregate", [
+          [{ target: dataStoreAddress, callData: dataStoreData }],
+        ]);
+
+        const body = {
+          jsonrpc: "2.0",
+          id: 1,
+          method: "eth_call",
+          params: [
+            {
+              to: multicallAddress,
+              data: multicallData,
+            },
+            "latest",
+          ],
+        };
+
+        try {
+          const response = await fetch(providerUrl, {
+            method: "POST",
+            headers: {
+              "Content-Type": "application/json",
+            },
+            body: JSON.stringify(body),
+            signal: controller.signal,
+          });
+
+          if (!response.ok) {
+            throw new Error(`Network response was not ok: ${response.statusText}`);
+          }
+
+          const { result } = await response.json();
+
+          if (result) {
+            const multicallResult = multicall.interface.decodeFunctionResult("blockAndAggregate", result);
+            const [sampleFieldValue] = dataStore.interface.decodeFunctionResult(
+              "getUint",
+              multicallResult.returnData[0].returnData
+            );
+
+            blockNumber = Number(multicallResult.blockNumber);
+            isSuccess = sampleFieldValue && sampleFieldValue > 0;
+          }
+        } catch (error) {
+          if (error.name !== "AbortError") {
+            throw error;
+          }
+        }
+      }
+
+      return {
+        url: providerUrl,
+        responseTime: Date.now() - startTime,
```

Comment:
not critical, but might be better to measure time of the http requests specifically, not the whole function call that e.g. includes some encoding/decoding, that might take some extra time and make the measurement a bit inaccurate

Reply:
i believe there are cases when it takes more time on slower machines? what is the downside of measuring time of request specifically?